src/Http/BaseApplication.php

```diff
@@ -61,6 +78,79 @@ public function __construct($configDir, EventManagerInterface $eventManager = nu
      */
     abstract public function middleware($middleware);
 
+    /**
+     * {@inheritDoc}
+     */
+    public function pluginEvents()
+    {
+        $events = $this->getEventManager();
+
+        foreach ($this->plugins->with('events') as $plugin) {
+            $events = $plugin->events($events);
+        }
+        $this->setEventManager($events);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function pluginMiddleware($middleware)
+    {
+        foreach ($this->plugins->with('middleware') as $plugin) {
+            $middleware = $plugin->middleware($middleware);
+        }
+
+        return $middleware;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function addPlugin($name, array $config = [])
+    {
+        if (is_string($name)) {
+            $plugin = $this->makePlugin($name, $config);
+        } else {
+            $plugin = $name;
+        }
+        $this->plugins->add($plugin);
+
+        return $this;
+    }
+
+    /**
+     * Get the plugin collection in use.
+     *
+     * @return \Cake\Core\PluginCollection
+     */
+    public function getPlugins()
+    {
+        return $this->plugins;
+    }
+
+    /**
+     * Create a plugin instance from a classname and configuration
+     *
+     * @param string $name The plugin classname
+     * @param array $config Configuration options for the plugin
+     * @return \Cake\Core\PluginInterface
+     */
+    public function makePlugin($name, array $config)
+    {
+        if (!class_exists($name)) {
```

Comment:
Could just check if name does not have backlash instead of class_exists(), similar to what App::className() does.